Fix for #21984: Viewing tables in Data Model is registered in "Recently viewed"

calherries opened the pull request:
Fixes https://github.com/metabase/metabase/issues/21984

### Root cause

The issue is that any GET requests to `/api/table/:id` will trigger a `:table-read` event, which records a view of the table, even if the requests come from the admin panel. Every request to that endpoint counts as a table view.

[This is](https://github.com/metabase/metabase/pull/17582/commits/a7a5ae37a967c472302b3aa8d20edd6b196bbed5) offending commit. It was part of the wider [Add recent views list PR](https://github.com/metabase/metabase/pull/17582) to add recently viewed items in the search bar.

### Proposed solution
I've added a `ignore_view=true` query parameter to all the places in the admin panel where components call `Tables.load(...)`, which tells the BE handler to skip triggering a `:table-read` event.

Commit:
Add explicit ignore_true parameter to Tables.load requests in admin panel to avoid counting the requests as a table view

Commit:
Remove .skip for repro

calherries commented:
@alxnddr @howonlee I'd appreciate a review on this since you worked on the PR that introduced recent views ([here](https://github.com/metabase/metabase/pull/17582)). 

Specifically - can you help track down where else GET `/api/table/:id` gets requested outside of calls from `Tables.load`? If there are no more calls outside `Tables.load`, this could be made a lot simpler, and we could just remove the :table-read event from the GET handler on the BE. i.e. [this line](https://github.com/metabase/metabase/blob/d4574e0b5fe409aad1d121319db3958edd5cf925/src/metabase/api/table.clj#L51).

Commit:
Fix formatting

howonlee commented:
IIRC there's actual third party usage of the API. This wiki page is not for no reason (https://github.com/metabase/metabase/wiki/Using-the-REST-API)

calherries commented:
@howonlee What are you suggesting though? I'm not changing the behaviour of existing callers here.

howonlee reviewed (approved):
This seems to work and fix the bug, I won't stop you from landing it w/ FE approval ~but if you change the API for third party use there also needs to be a small change in the docs for it, see here https://github.com/metabase/metabase/blob/master/docs/api/table.md. Not that we're perfect about this, really~

I completely forgot that we autogen those docs entirely, so they'll show up.

alxnddr reviewed (commented):
Seems to be working nicely! I'm wondering which api route tracks the view event when I open a table through the browse page? I don't see a request to `/api/table/:id` there. I'm asking because we might want to make it the opposite and make the `track_view` parameter instead since all occurences now have `ignore_view`

calherries commented:
@alxnddr Yes, that was exactly what I was hoping you'd spot by reviewing this. I also couldn't find any `/api/table/:id` requests that need a view tracked on the FE, but I was hoping you would confirm that. We could do as you suggested and add a `track_view` parameter and make the default behaviour to not track the view, OR remove [this line](https://github.com/metabase/metabase/blob/d4574e0b5fe409aad1d121319db3958edd5cf925/src/metabase/api/table.clj#L51) entirely. I would be in favour of the latter. Tables view counts already increase only when users view the data in the table or use it in a question/query, so counting any requests that only get the table schema shouldn't count as a "view" IMO.

calherries commented:
@alxnddr I just found another GET request to `/api/table/:id` [here](https://github.com/metabase/metabase/blob/master/enterprise/frontend/src/metabase-enterprise/audit_app/pages/AuditTableDetail.jsx#L17) via an `EntityName` component, which is a wrapper around `EntityObjectLoader` that just displays the table name. This realisation makes this PR's current solution likely to be incomplete. There are likely to be more places the `/api/table/:id` endpoint is used if any EntityObjectLoader can use it.

Is there a way we can pass an argument to an `EntityObjectLoader` like `ignore_view=true` to get the backend to disable counting the view when an EntityObjectLoader is used? If not, it's an argument in favour of disabling tracking views for all entity load requests. Ideally we would find a common solution for this and https://github.com/metabase/metabase/issues/18439, which has the same issue as this but for cards.

calherries commented:
More instances of EntityObjectLoaders that load tables: [TableLink](https://github.com/metabase/metabase/blob/dev-test/frontend/src/metabase/search/components/InfoText.jsx#L113) and [FieldAndTableName](https://github.com/metabase/metabase/blob/dev-test/frontend/src/metabase/parameters/components/ParameterSidebar.jsx#L253)

alxnddr commented:
> @alxnddr Yes, that was exactly what I was hoping you'd spot by reviewing this. I also couldn't find any `/api/table/:id` requests that need a view tracked on the FE, but I was hoping you would confirm that. We could do as you suggested and add a `track_view` parameter and make the default behaviour to not track the view, OR simply remove [this line](https://github.com/metabase/metabase/blob/d4574e0b5fe409aad1d121319db3958edd5cf925/src/metabase/api/table.clj#L51) entirely, disabling tracking views at this endpoint. I would be in favour of the latter. Tables view counts already increment when users view the data in the table or use it in a question/query, so counting any requests that only get the table schema or description etc without data shouldn't count as a "view" IMO.

Yeah, then it would be great just to remove tracking on `/api/table/:id` request. In this case we won't also need to add any parameters `EntityObjectLoaders` because now it seems like we want to have this behavior to be default and adding the parameter to 100% of usages.  
I'm not really good at figuring out Clojure code, did you find what exactly tracks opened table in the query builder?

nemanjaglumac committed:
Adapt repro to the fix

Repro was written before I know what the fix would look like.

We're conditionally choosing whether to display `HomePopularSection` or `HomeRecentSection`.
With this fix applied, data model table visit is not registered any more so there is nothing else really
that the test user visited. Hence, "Pick where you left off" is never displayed on the page and we
display `HomePopularSection` instead.

This adaption reflects those changes.

calherries commented:
Only other place a table-read event is [here](https://github.com/metabase/metabase/blob/21984-viewing-tables-in-admin-recently-viewed/src/metabase/api/dataset.clj#L53), when a user runs a `api/dataset` request that only includes that table. If the question has aggregations and filters it still counts as a view, but no joins are allowed. If the question is saved, running the question again will not count as a view. This seems reasonable as it counts a table view when a user starts a question and selects the table to begin querying from. But it excludes runs of saved queries such as those that would occur in a dashboard, which don't involve the user deliberately selecting a table, and shouldn't count as table views.

Commit:
Revert "Add explicit ignore_true parameter to Tables.load requests in admin panel to avoid counting the requests as a table view"

This reverts commit f887d0d2a6076c7b5073bf4356f3aa2361372055.

Commit:
Remove :table-read event from GET api/table/:id endpoint

Commit:
Remove unused ns

calherries commented:
Howon approved on slack, so I'm merging this
